### PR TITLE
BUG: Error while saving DataFrame with TimedeltaIndex to .csv #10833

### PR DESCRIPTION
**Pull request by soupault**
Fix in accordance with https://github.com/pydata/pandas/issues/10833

### CODE REVIEW

**Comment by jreback** [comment]
needs some tests. use the example you gave. add into ``test_frame.py``. needs a round-trip tests. Note that you will have to read it as ``object`` dtype and the run it thru ``to_timedelta`` on the expected frame.

**Comment by soupault** [comment]
It would be nice to refactor `test_frame.py` code... Btw, there are two existing assert methods: `self.assertXXX` and `assert_frame_equal`. I've used the first one as it seems like the second is to be reworked.

File: pandas/tests/test_frame.py [hunk 1]
```diff
@@ -6320,6 +6319,20 @@ def test_to_csv_from_csv(self):
...
+            df_orig = pd.DataFrame({'data': list(range(10))},
+                                   index=[i*dt for i in range(10)])
+            df_orig.index.rename('timestamp', inplace=True)
+            df_orig.to_csv(path)
+
+            df_test = pd.read_csv(path, index_col='timestamp')
+            df_test.index = pd.to_timedelta(df_test.index)
+            df_test.index.rename('timestamp', inplace=True)
+
+            self.assertTrue(df_test.equal(df_orig))
```
Comment (jreback) [inline pandas/tests/test_frame.py]: you *could* do this, but it is not nearly as imformative as ``assert_frame_equal``. pls change.

File: pandas/tests/test_frame.py [hunk 2]
```diff
@@ -6320,6 +6319,20 @@ def test_to_csv_from_csv(self):
             xp.columns = lmap(int,xp.columns)
             assert_frame_equal(xp,rs)
 
+        with ensure_clean() as path:
+            # GH 10833 (TimedeltaIndex formatting)
+            dt = pd.Timedelta(seconds=1)
+            df_orig = pd.DataFrame({'data': list(range(10))},
+                                   index=[i*dt for i in range(10)])
```
Comment (jreback) [inline pandas/tests/test_frame.py]: do ``index=Index([.........],name='timestamp')``

File: pandas/tests/test_frame.py [hunk 3]
```diff
@@ -6320,6 +6319,20 @@ def test_to_csv_from_csv(self):
...
+            df_orig.index.rename('timestamp', inplace=True)
```
Comment (jreback) [inline pandas/tests/test_frame.py]: we in general do not use ``inplace`` ops in testing code (nor recommend them for production code)

File: pandas/tests/test_frame.py [hunk 4]
```diff
@@ -6320,6 +6319,20 @@ def test_to_csv_from_csv(self):
...
+            df_orig.to_csv(path)
+
+            df_test = pd.read_csv(path, index_col='timestamp')
+            df_test.index = pd.to_timedelta(df_test.index)
```
Comment (jreback) [inline pandas/tests/test_frame.py]: call this ``result``

File: pandas/tests/test_frame.py [hunk 5]
```diff
@@ -6320,6 +6319,20 @@ def test_to_csv_from_csv(self):
             xp.columns = lmap(int,xp.columns)
             assert_frame_equal(xp,rs)
 
+        with ensure_clean() as path:
+            # GH 10833 (TimedeltaIndex formatting)
+            dt = pd.Timedelta(seconds=1)
+            df_orig = pd.DataFrame({'data': list(range(10))},
```
Comment (jreback) [inline pandas/tests/test_frame.py]: call this ``df``

File: pandas/tests/test_frame.py [hunk 6]
```diff
@@ -6320,6 +6319,20 @@ def test_to_csv_from_csv(self):
             xp.columns = lmap(int,xp.columns)
             assert_frame_equal(xp,rs)
 
+        with ensure_clean() as path:
+            # GH 10833 (TimedeltaIndex formatting)
+            dt = pd.Timedelta(seconds=1)
```
Comment (jreback) [inline pandas/tests/test_frame.py]: add a ``Timedelta`` column as well.

**Comment by jreback** [comment]
@soupault yes of course ``test_frame`` could use a refactor. but that's another PR!

File: pandas/tests/test_frame.py (hunk 3 above)
Comment (soupault) [inline pandas/tests/test_frame.py]: It seems like there is no way to preserve index name using `to_timedelta`.

    df_test.index = pd.to_timedelta(df_test.index)

Should I use this variant instead?

    df_test.index = pd.TimedeltaIndex(df_test.index, name='timestamp')

Comment (jreback) [inline pandas/tests/test_frame.py]: This works, no?

```
In [18]: s = Series(range(3),index=Index(timedelta_range('1 day',periods=3),name='foo'))

In [19]: s
Out[19]: 
foo
1 days    0
2 days    1
3 days    2
dtype: int64

In [20]: pd.to_timedelta(s.index)
Out[20]: TimedeltaIndex(['1 days', '2 days', '3 days'], dtype='timedelta64[ns]', name=u'foo', freq=None)
```

Comment (soupault) [inline pandas/tests/test_frame.py]: I have the same result as you on your piece of code.

Take a loot at this example:

    path = 'tempfile'
    dt = pd.Timedelta(seconds=1)
    df = pd.DataFrame({'dt_data': [i*dt for i in range(3)]},
                      index=pd.Index([i*dt for i in range(3)],
                                  name='dt_index'))
    df.to_csv(path)

    result = pd.read_csv(path, index_col='dt_index')
    print(result.index)
    >>> Index(['00:00:00', '00:00:01', '00:00:02'], dtype='object', name='dt_index')
    result.index = pd.to_timedelta(result.index)
    print(result.index)
    >>> TimedeltaIndex(['00:00:00', '00:00:01', '00:00:02'], dtype='timedelta64[ns]', freq=None)

Seems like it disappears after `Index` -> `TimedeltaIndex` conversion.

Comment (jreback) [inline pandas/tests/test_frame.py]: oh i c. if you want to trace where that is not passed thru would be gr8. (if its easy do it as part of this issue, if not, pls open a new issue).

Comment (soupault) [inline pandas/tests/test_frame.py]: ok, I'll finalize both issues within this week.

Comment (jreback) [inline pandas/tests/test_frame.py]: gr8 thanks. Add a separate whatsnew note (with same issue number) for the name lost in conversion

Comment (soupault) [inline pandas/tests/test_frame.py]: I'd say that casting `to_datetime` on index doesn't keep the name also. Let's open a new issue.
This PR can't be merged before I fix the new one.

File: pandas/tests/test_frame.py [hunk 7]
```diff
@@ -6320,6 +6319,20 @@ def test_to_csv_from_csv(self):
...
+            df = pd.DataFrame({'dt_data': [i*dt for i in range(3)]},
+                              index=pd.Index([i*dt for i in range(3)],
+                                             name='dt_index'))
+            df.to_csv(path)
+
+            result = pd.read_csv(path, index_col='dt_index')
+            result.index = pd.to_timedelta(result.index)
+            result['dt_data'] = pd.to_timedelta(result['dt_data'])
+
```
Comment (jreback) [inline pandas/tests/test_frame.py]: set the index name for now so this issue is independent

File: pandas/tests/test_frame.py [hunk 8]
```diff
@@ -6320,6 +6319,22 @@ def test_to_csv_from_csv(self):
...
+            # GH 10833 (TimedeltaIndex formatting)
+            dt = pd.Timedelta(seconds=1)
+            df = pd.DataFrame({'dt_data': [i*dt for i in range(3)]},
+                              index=pd.Index([i*dt for i in range(3)],
+                                             name='dt_index'))
+            df.to_csv(path)
+
+            result = pd.read_csv(path, index_col='dt_index')
+            result.index = pd.to_timedelta(result.index)
+            # TODO: remove renaming when GH 10875 is solved
```
Comment (jreback) [inline pandas/tests/test_frame.py]: fine, though ``result.index.name = 'dt_index'`` as well

**Comment by jreback** [comment]
thank you sir!

**Pull request merged**